libr/debug/debug.c

```diff
@@ -1145,6 +1105,45 @@ R_API int r_debug_continue_kill(RDebug *dbg, int sig) {
 	if (!dbg) {
 		return 0;
 	}
+
+	if (dbg->session) {
+		// If the debugger is not at the end of the session
+		// Go to the end or the next breakpoint
+		if (dbg->session->cnum != dbg->session->maxcnum) {
+			bool has_bp = false;
+			RRegItem *ripc = r_reg_get (dbg->reg, dbg->reg->name[R_REG_NAME_PC], R_REG_TYPE_GPR);
+			RVector *vreg = ht_up_find (dbg->session->registers, ripc->offset | (ripc->arena << 16), NULL);
+			RDebugChangeReg *reg;
+			r_vector_foreach_prev (vreg, reg) {
+				if (reg->cnum <= dbg->session->cnum) {
+					continue;
+				}
+				has_bp = r_bp_get_in (dbg->bp, reg->data, R_BP_PROT_EXEC) != NULL;
+				if (has_bp) {
+					eprintf ("hit breakpoint at: 0x%" PFMT64x " cnum: %d\n", reg->data, reg->cnum);
+					r_debug_goto_cnum (dbg, reg->cnum);
+					return dbg->tid;
+				}
+			}
+
+			r_debug_goto_cnum (dbg, dbg->session->maxcnum);
+			return dbg->tid;
+		}
+
+		if (dbg->trace_continue) {
+			// Continuously stepping to trace each instruction until user interrupt ^C
+			r_cons_break_push (NULL, NULL);
+			while (!r_cons_is_breaked ()) {
+				r_debug_step (dbg, 1);
+				if (dbg->session->reasontype != R_DEBUG_REASON_STEP) {
```

Comment:
While testing I have attempted to `dc` to main in trace mode and I kept being stopped on the way there because reasontype was R_DEBUG_REASON_BREAKPOINT (while not stepping on any breakpoint on the PC).
This should be looked into. Maybe the session section should be implemented within continue's "repeat" loop below and use it's existing logic to determine whether `dc` is finished instead of re-implementing it.

Reply:
I have encountered this which is caused by syscall trap because PTRACE_O_TRACESYSGOOD is on by default when attached. You are right and thanks for the suggestion. I will them move into the existing loop.

Reply:
Then, user has to set `e dbg.bpsysign=true` to ignore those system breakpoints which is already implemented in the loop.